src/cudamatrix/cu-matrix.cc

```diff
@@ -2540,33 +2572,53 @@ void CuMatrixBase<Real>::Lookup(const std::vector<Int32Pair> &indices,
   // Checks the pointer.
   KALDI_ASSERT(output != NULL);
 
-  // Resizes the output vector.
-  output->resize(indices.size());
-
 #if HAVE_CUDA == 1
   if (CuDevice::Instantiate().Enabled()) {
     CuArray<Int32Pair> cuda_indices(indices);
-    CuArray<Real> cuda_output(output->size());
+    Lookup(cuda_indices, output);
+  } else
+#endif
+  {
+    for (int32 i = 0; i < indices.size(); i++) {
+      output[i] = (*this)(indices[i].first, indices[i].second);
+    }
+  }
+}
+
+template<typename Real>
+void CuMatrixBase<Real>::Lookup(const CuArray<Int32Pair> &indices,
+				Real *output) const {
+  int32 num_elements = indices.Dim();
+  if (num_elements == 0) return;
+  // Checks the pointer.
+  KALDI_ASSERT(output != NULL);
 
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    CuArray<Real> cuda_output(num_elements);
     Timer tim;
     dim3 dimBlock(CU1DBLOCK, 1);
-    dim3 dimGrid(n_blocks(indices.size(), CU1DBLOCK), 1);
-    
+    dim3 dimGrid(n_blocks(num_elements, CU1DBLOCK), 1);
+
     cuda_matrix_lookup(dimGrid, dimBlock, this->data_, this->Dim(),
-                       cuda_indices.Data(), indices.size(), cuda_output.Data());
+                       indices.Data(), num_elements, cuda_output.Data());
     CU_SAFE_CALL(cudaGetLastError());
-
-    cuda_output.CopyToVec(output);
     
+    cuda_output.CopyToHost(output);
     CuDevice::Instantiate().AccuProfile(__func__, tim.Elapsed());
   } else
 #endif
   {
-    for (int32 i = 0; i < indices.size(); i++) {
-      (*output)[i] = (*this)(indices[i].first, indices[i].second);
+    MatrixIndexT num_rows = this->num_rows_, num_cols = this->num_cols_;
+    for (int32 i = 0; i < num_elements; i++) {
+      Int32Pair index = indices.Data()[i];
```

Comment:
Here and in similar code elsewhere, it might be best to avoid the function call to Data() each time, by declaring a pointer outside the loop.